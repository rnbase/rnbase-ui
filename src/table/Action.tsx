import React from 'react';
import {
  StyleProp,
  Image,
  ImageSourcePropType,
  ImageStyle,
  StyleSheet,
  Text,
  TextProps,
  View,
} from 'react-native';

import { Theme, Themed, withTheme } from '../theming';

interface ActionProps extends TextProps {
  children?: React.ReactNode;
  iconSource?: ImageSourcePropType;
  iconStyle?: StyleProp<ImageStyle>;
}

const Action: React.FC<Themed<typeof createStyleSheet, ActionProps>> = ({
  theme: { styles },
  children,
  style,
  iconSource,
  iconStyle,
  ...props
}) => (
  <View style={styles.root}>
    {children && (
      <Text {...props} style={[styles.text, style]} numberOfLines={1}>
        {children}
      </Text>
    )}
    {!!iconSource && <Image style={[styles.icon, iconStyle]} source={iconSource} />}
  </View>
);

const createStyleSheet = ({ colors, fonts }: Theme) =>
  StyleSheet.create({
    root: {
      alignItems: 'center',
      flexDirection: 'row',
    },
    text: {
      ...fonts.normal,
      height: 16,
      fontSize: 16,
      lineHeight: 16,
      color: colors.blue,
    },
    icon: {
      width: 22,
      height: 22,
      marginRight: -4,
      tintColor: colors.blue,
    },
  });

export default withTheme(Action, createStyleSheet, 'Table.Action');
