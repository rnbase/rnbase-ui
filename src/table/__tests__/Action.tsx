import React from 'react';

// Note: test renderer must be required after react-native.
import TestRenderer from 'react-test-renderer';

import Action from '../Action';

type Props = React.ComponentProps<typeof Action>;

/**
 * Under test
 */
const createElement = (props: Partial<Props>) => <Action {...props}>text</Action>;

const createRenderer = (props: Partial<Props>) => TestRenderer.create(createElement(props));

const iconSource = {
  uri: '../../assets/star.png',
};

it('should render normally', () => {
  expect(createRenderer({})).toMatchSnapshot();
});

it('should render with custom styles', () => {
  expect(createRenderer({ style: { color: 'red' } })).toMatchSnapshot();
});

it('should render with icon', () => {
  expect(createRenderer({ iconSource })).toMatchSnapshot();
});

it('should render with icon and custom icon styles', () => {
  expect(createRenderer({ iconSource, iconStyle: { tintColor: 'red' } })).toMatchSnapshot();
});
