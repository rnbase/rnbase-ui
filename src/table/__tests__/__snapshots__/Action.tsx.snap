// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render normally 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
    }
  }
>
  <Text
    numberOfLines={1}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "color": "#007aff",
          "fontFamily": "System",
          "fontSize": 16,
          "fontWeight": "400",
          "height": 16,
          "includeFontPadding": false,
          "lineHeight": 16,
          "textAlignVertical": "center",
        },
        undefined,
      ]
    }
  >
    text
  </Text>
</View>
`;

exports[`should render with custom styles 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
    }
  }
>
  <Text
    numberOfLines={1}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "color": "#007aff",
          "fontFamily": "System",
          "fontSize": 16,
          "fontWeight": "400",
          "height": 16,
          "includeFontPadding": false,
          "lineHeight": 16,
          "textAlignVertical": "center",
        },
        Object {
          "color": "red",
        },
      ]
    }
  >
    text
  </Text>
</View>
`;

exports[`should render with icon 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
    }
  }
>
  <Text
    numberOfLines={1}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "color": "#007aff",
          "fontFamily": "System",
          "fontSize": 16,
          "fontWeight": "400",
          "height": 16,
          "includeFontPadding": false,
          "lineHeight": 16,
          "textAlignVertical": "center",
        },
        undefined,
      ]
    }
  >
    text
  </Text>
  <Image
    source={
      Object {
        "uri": "../../assets/star.png",
      }
    }
    style={
      Array [
        Object {
          "height": 22,
          "marginRight": -4,
          "tintColor": "#007aff",
          "width": 22,
        },
        undefined,
      ]
    }
  />
</View>
`;

exports[`should render with icon and custom icon styles 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
    }
  }
>
  <Text
    numberOfLines={1}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "color": "#007aff",
          "fontFamily": "System",
          "fontSize": 16,
          "fontWeight": "400",
          "height": 16,
          "includeFontPadding": false,
          "lineHeight": 16,
          "textAlignVertical": "center",
        },
        undefined,
      ]
    }
  >
    text
  </Text>
  <Image
    source={
      Object {
        "uri": "../../assets/star.png",
      }
    }
    style={
      Array [
        Object {
          "height": 22,
          "marginRight": -4,
          "tintColor": "#007aff",
          "width": 22,
        },
        Object {
          "tintColor": "red",
        },
      ]
    }
  />
</View>
`;
